## Load plink before starting R
# module load plink/1.90b6.6
# R

## Now run R code
library('data.table')
library('SummarizedExperiment')
library('sessioninfo')

## To avoid issues with running this code on qsub
data.table::setDTthreads(threads = 1)

## Find the samples for this project
load("/dcl01/lieber/ajaffe/lab/Nicotine/NAc/RNAseq/paired_end_n239/count_data/NAc_Nicotine_hg38_rseGene_rawCounts_allSamples_n239.rda", verbose = TRUE)
stopifnot(length(unique(rse_gene$BrNum)) == ncol(rse_gene))

## For converting BrNum's into numbers
brnumerical <- function(x) {
    as.integer(gsub("Br|_.*", "", x))
}

libd_bfile <- "/dcl01/lieber/ajaffe/Brain/Imputation/Subj_Cleaned/LIBD_merged_h650_1M_Omni5M_Onmi2pt5_Macrogen_QuadsPlus_dropBrains_maf01_hwe6_geno10_hg38"

## Read the LIBD fam data
libd_fam <- fread(
    paste0(libd_bfile, ".fam"),
    col.names = c('famid', 'w_famid', 'w_famid_fa', 'w_famid_mo', 'sex_code', 'phenotype'))
libd_fam$brnumerical <- brnumerical(libd_fam$famid)
setkey(libd_fam, 'brnumerical')


## Filter the LIBD data to the one specific to this project
# region <- "NAc_Nicotine"
message(paste(Sys.time(), 'processing', "NAc_Nicotine"))
samp_file <- paste0('samples_to_extract_', "NAc_Nicotine", '.txt')

fwrite(
    libd_fam[brnumerical(rse_gene$BrNum), 1:2], ## can be more involved
    file = samp_file,
    sep = '\t', col.names = FALSE
)
newbfile <- paste0(
    '/dcl01/lieber/ajaffe/Matt/MNT_thesis/snRNAseq/10x_pilot_FINAL/twas/filter_data/duplicate_snps_bim/LIBD_merged_h650_1M_Omni5M_Onmi2pt5_Macrogen_QuadsPlus_dropBrains_maf01_hwe6_geno10_hg38_filtered_',
    "NAc_Nicotine_duplicateSNPs"
)

## Extract
message(paste(Sys.time(), 'running bfile extract'))

system(paste("plink --bfile", libd_bfile,
	"--keep", samp_file, "--make-bed --out",
	newbfile, " --memory 100000 --biallelic-only"))

# message(paste(Sys.time(), "reading the bim file", newbfile))
# bim <- fread(
#     paste0(newbfile, ".bim"),
#     col.names = c("chr", "snp", "position", "basepair", "allele1", "allele2")
# )
#
# # > table(duplicated(bim$snp))
# #
# #    FALSE     TRUE
# # 10943065    44114
#
# newsnp <- make.names(bim$snp, unique = T)
# bim$snp <- newsnp
#
# fwrite(bim, file = paste0(newbfile, ".bim"), sep = " ", col.names = FALSE)


## Reproducibility information
print('Reproducibility information:')
Sys.time()
proc.time()
options(width = 120)
session_info()
# ??? Session info ?????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????
#  setting  value
#  version  R version 4.0.2 Patched (2020-06-24 r78746)
#  os       CentOS Linux 7 (Core)
#  system   x86_64, linux-gnu
#  ui       X11
#  language (EN)
#  collate  en_US.UTF-8
#  ctype    en_US.UTF-8
#  tz       US/Eastern
#  date     2020-08-27
#
# ??? Packages ?????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????
#  package              * version  date       lib source
#  assertthat             0.2.1    2019-03-21 [2] CRAN (R 4.0.0)
#  Biobase              * 2.48.0   2020-04-27 [2] Bioconductor
#  BiocGenerics         * 0.34.0   2020-04-27 [2] Bioconductor
#  bitops                 1.0-6    2013-08-17 [2] CRAN (R 4.0.0)
#  cli                    2.0.2    2020-02-28 [2] CRAN (R 4.0.0)
#  crayon                 1.3.4    2017-09-16 [2] CRAN (R 4.0.0)
#  data.table           * 1.12.8   2019-12-09 [2] CRAN (R 4.0.0)
#  DelayedArray         * 0.14.0   2020-04-27 [2] Bioconductor
#  fansi                  0.4.1    2020-01-08 [2] CRAN (R 4.0.0)
#  GenomeInfoDb         * 1.24.0   2020-04-27 [2] Bioconductor
#  GenomeInfoDbData       1.2.3    2020-05-18 [2] Bioconductor
#  GenomicRanges        * 1.40.0   2020-04-27 [2] Bioconductor
#  glue                   1.4.1    2020-05-13 [2] CRAN (R 4.0.0)
#  IRanges              * 2.22.1   2020-04-28 [2] Bioconductor
#  lattice                0.20-41  2020-04-02 [3] CRAN (R 4.0.2)
#  Matrix                 1.2-18   2019-11-27 [3] CRAN (R 4.0.2)
#  matrixStats          * 0.56.0   2020-03-13 [2] CRAN (R 4.0.0)
#  RCurl                  1.98-1.2 2020-04-18 [2] CRAN (R 4.0.0)
#  S4Vectors            * 0.26.1   2020-05-16 [2] Bioconductor
#  sessioninfo          * 1.1.1    2018-11-05 [2] CRAN (R 4.0.0)
#  SummarizedExperiment * 1.18.1   2020-04-30 [2] Bioconductor
#  withr                  2.2.0    2020-04-20 [2] CRAN (R 4.0.0)
#  XVector                0.28.0   2020-04-27 [2] Bioconductor
#  zlibbioc               1.34.0   2020-04-27 [2] Bioconductor
#
# [1] /users/aseyedia/R/4.0
# [2] /jhpce/shared/jhpce/core/conda/miniconda3-4.6.14/envs/svnR-4.0/R/4.0/lib64/R/site-library
# [3] /jhpce/shared/jhpce/core/conda/miniconda3-4.6.14/envs/svnR-4.0/R/4.0/lib64/R/library
system('plink --version')
# PLINK v1.90b6.6 64-bit (10 Oct 2018)
